### ISSUE TITLE
Difference in the functions `from_hdu_list()` and `from_hdulist()`

### ISSUE THREAD

**Issue by AtreyeeS** [gammapy/gammapy#1453]
Hello,

I am a bit confused by the difference in the two functions

`from_hdu_list()` in `class Map` (base.py) and `from_hdulist()` in `class MapGeom` (geom.py)

I can see that `from_hdu_list()` is used to properly call `from_hdulist()` , but maybe it is a bit confusing to an user to see 2 very closely named functions callable from a map object?

There is also a similar  `from_hdulist()` in `class HpxMap`.

**Comment by cdeil** [comment]
For reference: see code in [Map.from_hdu_list](http://docs.gammapy.org/dev/_modules/gammapy/maps/base.html#Map.from_hdu_list) and [MapGeom.from_hdulist](http://docs.gammapy.org/dev/_modules/gammapy/maps/geom.html#MapGeom.from_hdulist). Also see other classes with [from_hdulist](http://docs.gammapy.org/dev/search.html?q=from_hdulist&check_keywords=yes&area=default#) or [from_hdu_list](http://docs.gammapy.org/dev/search.html?q=from_hdu_list&check_keywords=yes&area=default#) methods.

The idea is to have methods to allow FITS serialisation on most Gammapy objects. To give flexibility, in addition to read / write that hit the disk, there's extra methods which just serialise / deserialise to `astropy.io.fits` objects, usually an HDU or HDUList. 

In addition, in `gammapy.maps` there is a class hierarchy, and there are some "factory methods" on the base classes (like `Map.read`) that are conveniences and identify and dispatch to subclass constructors.

So in principle it's normal that there are many such methods, and e.g. `MapGeom` can serialise itself to FITS and then Map calls the geom FITS serialisation code as part of it's FITS serialisation. Although probably it could also have been done differently, by putting everything in the Map methods, given that it's rare to only want to serialise a geom without data. @woodmd could explain the design.

@AtreyeeS - Could you have another look and comment a bit more specifically which method you think  should be removed or changed or clarified what it does in the docstring?

One obvious point is that we should pick either `hdu_list` or `hdulist` and use it consistently in Gammapy, instead of the current mix. @woodmd - do you think it's still possible to make the change in Gammapy v0.8 for gammapy.maps, or is it inacceptable because there's too many callers in Fermipy already? (I think for our users, it would be OK, people are just starting to use gammapy.maps now and we will make the v0.8 release soon).

**Comment by cdeil** [comment]
I also think we should significantly extend the Gammapy.maps docs concerning FITS I/O.
This is what @woodmd wrote: http://docs.gammapy.org/dev/maps/index.html#fits-i-o
and @AtreyeeS just added [intro_maps.ipynb](https://github.com/gammapy/gammapy-extra/blob/master/notebooks/intro_maps.ipynb)

There are many FITS example files here:
http://gamma-astro-data-formats.readthedocs.io/en/latest/skymaps/index.html

The question is how to grow the gammapy.maps docs. E.g. one could make a separate notebook showing examples of FITS serialisation, or one could split that content by WCS and HPX and add FITS examples to the sub-pages: http://docs.gammapy.org/dev/maps/index.html#using-gammapy-maps

@AtreyeeS @woodmd @registerrier  - Thoughts?

**Comment by mackaiver** [comment]
> do you think it's still possible to make the change in Gammapy v0.8 for gammapy.maps, or is it inacceptable because there's too many callers in Fermipy already?

As far as I can see in the FermiPy repo, there are no  callers for either `from_hdu_list` or `from_hdulist`.

**Comment by mackaiver** [comment]
So the base class `Map` has a `from_hdu_list` which calls `from_hdulist` from the implementing class. Sounds like a typo to me.

**Comment by AtreyeeS** [comment]
From what I understand, the `from_hdu_list`  function is used to get the Map class using `_get_map_cls` and call the according `from_hdulist`. But since the map object already knows what geom it has, maybe we can do away with the `from_hdu_list` function?

Or at the very least, make it private so that it doesn't show up in the function suggestions?
Or maybe we shift completely to one naming and thus avoid confusion!

**Comment by woodmd** [comment]
I don't think calling the method `from_hdu_list` instead of `from_hdulist` was intentional.  Since `from_hdulist` is used in the derived classes I agree it makes sense to use the same name for `Map`.  

However I woudn't recommend removing this method from `Map`.  The `Map.read` method (which relies on `Map.from_hdu_list`) is useful for querying the contents of a file when you don't necessarily know the type of map object it contains.

**Comment by mackaiver** [comment]
Alright. So I'll rename the method.

**Comment by cdeil** [comment]
I had a look at this.

Using a simple example:
```
class Map:
    @classmethod
    def read(cls):
        print('Map.read')
        cls.from_hdulist()
    @classmethod
    def from_hdulist(cls):
        print('Map.from_hdulist')

class SomeMap(Map):
    @classmethod
    def from_hdulist(cls):
        print('SomeMap.from_hdulist')
        
Map.read()
print('---')
SomeMap.read()
```
one finds
```
Map.read
Map.from_hdulist
---
Map.read
SomeMap.from_hdulist
```
i.e. `SomeMap.read` will have `cls=SomeMap` and then call `SomeMap.from_hdulist`.

But what we have is incompatible signature of `Map.from_hdulist` (with argument `map_type`) and derived map `from_hdulist` (without argument `map_type`)
```
gammapy/maps/wcsmap.py
96:    def from_hdulist(cls, hdu_list, hdu=None, hdu_bands=None):

gammapy/maps/hpxmap.py
98:    def from_hdulist(cls, hdu_list, hdu=None, hdu_bands=None):

gammapy/maps/base.py
167:            return Map.from_hdulist(hdulist, hdu, hdu_bands, map_type)
211:    def from_hdulist(hdulist, hdu=None, hdu_bands=None, map_type='auto'):
215:        return cls_out.from_hdulist(hdulist, hdu=hdu, hdu_bands=hdu_bands)
```
resulting in these errors:
https://travis-ci.org/gammapy/gammapy/jobs/402197541#L2092
```
E           TypeError: from_hdulist() takes from 2 to 4 positional arguments but 5 were given
gammapy/maps/base.py:167: TypeError
```

Looking at the implementation of the construct / read related methods on the `Map` base class, I think  one solution is to make them all `@staticmethod` instead of `@classmethod`, thus avoiding the mix of calling base and derived methods. This is a simple change after which all tests pass. I'll add a commit for that in #1469 now.

@woodmd, and anyone else that has time to review - please have a look at #1469!

**Comment by cdeil** [comment]
Should be resolved via #1469 .
